src/osd/OSD.cc

```diff
@@ -7954,22 +7954,28 @@ void OSD::trim_maps(epoch_t oldest, bool skip_maps)
   if (min <= superblock.oldest_map)
     return;
 
-  ObjectStore::Transaction t;
+  vector<ObjectStore::Transaction> tls;
+  tls.push_back(ObjectStore::Transaction());
   // Trim from the superblock's oldest_map up to `min`.
   // Break if we have exceeded the txn target size.
   // If skip_maps is true, we will trim up `min` unconditionally.
-  while ((superblock.oldest_map < min) &&
-         (skip_maps ||
-          t.get_num_ops() <= cct->_conf->osd_target_transaction_size)) {
+  while (superblock.oldest_map < min) {
     dout(20) << " removing old osdmap epoch " << superblock.oldest_map << dendl;
-    t.remove(coll_t::meta(), get_osdmap_pobject_name(superblock.oldest_map));
-    t.remove(coll_t::meta(), get_inc_osdmap_pobject_name(superblock.oldest_map));
+    tls.back().remove(coll_t::meta(), get_osdmap_pobject_name(superblock.oldest_map));
+    tls.back().remove(coll_t::meta(), get_inc_osdmap_pobject_name(superblock.oldest_map));
     ++superblock.oldest_map;
+    if (tls.back().get_num_ops() > cct->_conf->osd_target_transaction_size) {
+      if (skip_maps) {
+        tls.push_back(ObjectStore::Transaction());
+      } else {
+        break;
+      }
+    }
   }
-  if (t.get_num_ops() > 0) {
+  if (tls.front().get_num_ops() > 0) {
     service.publish_superblock(superblock);
-    write_superblock(cct, superblock, t);
-    int tr = store->queue_transaction(service.meta_ch, std::move(t), nullptr);
+    write_superblock(cct, superblock, tls.back());
+    int tr = store->queue_transactions(service.meta_ch, tls, nullptr);
```

Comment:
This does bound the size of any individual transaction, but the store is still required to commit entire vector of transactions atomically as though they were a single really big one.  What we actually want to do is bound the amount of work the store needs to do in a single atomic commit, so I think we really do want to retain the original behavior, but with a claim_and_reset() method instead of relying on the move constructor.

Reply:
> This does bound the size of any individual transaction, but the store is still required to commit entire vector of transactions atomically as though they were a single really big one.

I missed that part, fixed now.

> so I think we really do want to retain the original behavior, but with a claim_and_reset() method instead of relying on the move constructor.

I was able to retain the original behavior without introducing new state to Transaction. I think we should avoid holding a not-reset transaction instance if we can. Do find that plausible?